src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -100,56 +131,111 @@ fn execute(top_match: clap::ArgMatches) -> Result<(), String> {
-        directory.mut_directories().push({
-          let mut dir_node = bazel_protos::remote_execution::DirectoryNode::new();
-          dir_node.set_name(file_name_as_utf8(&path)?);
-          dir_node.set_digest(fingerprint_to_digest(&fingerprint, proto_size_bytes as i64));
-          dir_node
-        })
-      }
-      fs::Stat::File(fs::File {
-                       path,
-                       is_executable,
-                     }) => {
-        let (fingerprint, size_bytes) = save_file(store, &path)?;
-        directory.mut_files().push({
-          let mut file_node = bazel_protos::remote_execution::FileNode::new();
-          file_node.set_name(file_name_as_utf8(&path)?);
-          file_node.set_digest(fingerprint_to_digest(&fingerprint, size_bytes as i64));
-          let mut absolute_path = root.to_path_buf();
-          absolute_path.push(path.file_name().ok_or_else(|| {
-            format!("{:?} did not have a file_name", path)
-          })?);
-          file_node.set_is_executable(is_executable);
-          file_node
-        });
-      }
-      fs::Stat::Link(fs::Link(path)) => {
-        return Err(format!("Don't yet know how to handle symlinks: {:?}", path))
-      }
-    }
-  }
-  store.record_directory(&directory)
+fn save_file(
+  store: Arc<Store>,
+  posix_fs: &fs::PosixFS,
+  file: fs::File,
+) -> BoxFuture<(Fingerprint, usize), String> {
+  posix_fs
+    .read_file(&file)
+    .map_err(move |err| {
+      format!("Error reading file {:?}: {}", file, err.description())
+    })
+    .and_then(move |content| {
+      Ok((
+        store.store_file_bytes(&content.content)?,
+        content.content.len(),
+      ))
+    })
+    .to_boxed()
+}
+
+fn save_directory(
+  store: Arc<Store>,
+  posix_fs: Arc<fs::PosixFS>,
+  base_dir: Arc<fs::Dir>,
+) -> BoxFuture<(Fingerprint, usize), String> {
+  let store_copy = store.clone();
+  Box::new(
+    posix_fs
+      .scandir(&base_dir)
+      .map_err(move |err| {
+        format!("Failed to scandir {:?}: {}", &base_dir, err.description())
+      })
+      .and_then(move |stats| {
+        let mut dirs = Vec::new();
+        let mut files = Vec::new();
+        for stat in stats.into_iter() {
+          match stat {
+            fs::Stat::Dir(dir) => dirs.push(dir),
+            fs::Stat::File(file) => files.push(file),
+            fs::Stat::Link(fs::Link(path)) => {
+              let err: BoxFuture<_, _> = Box::new(futures::future::err(format!(
+                "Don't yet know how to handle symlinks: {:?}",
+                path
+              )));
+              return err;
+            }
+          }
+        }
+
+        let store_copy = store.clone();
+        let posix_fs_copy = posix_fs.clone();
+        let dir_futures: Vec<_> = dirs
+          .into_iter()
+          .map(move |dir| {
+            let dir = Arc::new(dir);
+            save_directory(store_copy.clone(), posix_fs_copy.clone(), dir.clone())
```

Comment:
As evidenced by the call to `store.clone()` with `file_to_file_node` below, it's possible that the `store_copy` alias is also not necessary?

Reply:
Alas not, and the new naming maybe helps make that clear.

store_inner_copy gets moved into the map call's closure.
store_inner_copy.clone() gets moved into save_directory call.
store_copy gets moved into the file_futures map call's closure.

All three need to exist :(